### PR TITLE
Issues API unit tests

### PR DESCRIPTION
**Pull request by BenFradet**
Still need to add test cases to `ApiSpec`

### CODE REVIEW

**Comment by BenFradet** [comment]
>The command "sbt ++$TRAVIS_SCALA_VERSION test:fastOptJS" exited with 126.
0.00s$ sbt ++$TRAVIS_SCALA_VERSION orgScriptCI
/home/travis/.travis/job_stages: line 724: /usr/local/bin/sbt: Permission denied

**Comment by fedefernandez** [comment]
I'm not sure what could be the cause. I'll try to take a look tomorrow.

**Comment by juanpedromoreno** [comment]
@BenFradet @fedefernandez This might solve the issues: https://github.com/47deg/github4s/pull/117

**Comment by BenFradet** [comment]
I still have a few unit tests failing, still investigating why

File: github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala [hunk 1]
```diff
@@ -596,4 +597,98 @@ trait MockGithubApiServer extends MockServerService with FakeResponses with Test
...
+        .withQueryStringParameters(new Parameter("q", "+"))
+        .withHeader("Authorization", tokenHeader))
+    .respond(response.withStatusCode(okStatusCode).withBody(searchIssuesValidResponse))
+
+  mockServer
+    .when(
+      request
+        .withMethod("GET")
+        .withPath(s"/search/issues")
+        .withQueryStringParameters(new Parameter("q", s"$nonExistentSearchQuery+"))
```
Comment (BenFradet) [inline github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala]: have you guys tried `withQueryStringParameters`  before?
Never seems to match.

File: github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala [hunk 2]
```diff
@@ -756,4 +757,145 @@ class ApiSpec
...
+      validRepoName,
+      validIssueTitle,
+      validIssueBody,
+      None,
+      List.empty,
+      List.empty)
+    response should be('left)
+  }
+
+  "Issues >> Edit" should "return the edited issue if valid data is provided" ignore {
```
Comment (BenFradet) [inline github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala]: for some reason this one's failing too

File: github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala (hunk 1 above)
Comment (BenFradet) [inline github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala]: fixed it with regex :+1:

File: github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala (hunk 2 above)
Comment (BenFradet) [inline github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala]: You guys seem to have used post for patch request for the git data API: https://github.com/47deg/github4s/blob/master/github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala#L328-L348.

So I'm assuming there is something wrong with how mock server deals with patch?

**Comment by BenFradet** [comment]
Good to review :+1:

File: github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala (hunk 2 above)
Comment (fedefernandez) [inline github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala]: We should create an issue to fix that part. The root of the problem is the call to [`postData`](https://github.com/47deg/github4s/blob/f6f11910343e9e958d60c9829113b2209e4f39bb/github4s/jvm/src/main/scala/github4s/HttpRequestBuilderExtensionJVM.scala#L58-L58). In that part the request already have set the `PATCH` method but if we take a look to this [method in scalaj-http](https://github.com/scalaj/scalaj-http/blob/master/src/main/scala/scalaj/http/Http.scala#L465) we can see that it's setting the method to `POST`.

One solution could be to replace the `postData` call by:
```diff
-- .postData(body)
++ .postData(d).method(rb.httpVerb.verb)
```

But I've not tested it yet. Sorry about the confusion

File: github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala [hunk 3]
```diff
@@ -596,4 +597,98 @@ trait MockGithubApiServer extends MockServerService with FakeResponses with Test
...
+        .withMethod("POST")
+        .withPath(s"/repos/$validRepoOwner/$validRepoName/issues")
+        .withHeader(not("Authorization")))
+    .respond(response.withStatusCode(unauthorizedStatusCode).withBody(unauthorizedResponse))
+
+  mockServer
+    .when(
+      request
+        .withMethod("POST")
+        .withPath(s"/repos/$validRepoOwner/$validRepoName/issues")
```
Comment (fedefernandez) [inline github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala]: Should it be `invalidRepoName`?

File: github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala [hunk 4]
```diff
@@ -596,4 +597,98 @@ trait MockGithubApiServer extends MockServerService with FakeResponses with Test
...
+      request
+        .withMethod("PATCH")
+        .withPath(s"/repos/$validRepoOwner/$validRepoName/issues/$validIssueNumber")
+        .withHeader(not("Authorization")))
+    .respond(response.withStatusCode(unauthorizedStatusCode).withBody(unauthorizedResponse))
+
+  mockServer
+    .when(
+      request
+        .withMethod("PATCH")
```
Comment (fedefernandez) [inline github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala]: I think that our http client is applying a `POST` method for this operation

File: github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala (hunk 1 above)
Comment (fedefernandez) [inline github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala]: 👍

File: github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala (hunk 3 above)
Comment (BenFradet) [inline github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala]: true, I don't know how the tests manage to pass :confused:

File: github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala (hunk 4 above)
Comment (BenFradet) [inline github4s/jvm/src/test/scala/github4s/utils/MockGithubApiServer.scala]: Yup I modified those, the sync doesn't seem to be reflected on gh, weird as well.

**Comment by BenFradet** [comment]
nvm, pushed to my remote :'(

File: github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala (hunk 2 above)
Comment (BenFradet) [inline github4s/jvm/src/test/scala/github4s/unit/ApiSpec.scala]: I get it, thanks. I'll log an issue :+1:

**Comment by BenFradet** [comment]
> The command "sbt ++$TRAVIS_SCALA_VERSION orgScriptCI" exited with 137.

on 2.12

**Comment by fedefernandez** [comment]
`/home/travis/.travis/job_stages: line 738:  5784 Killed`
😞 retry?

**Pull request merged**